### ISSUE TITLE
the lifecycle of GreenMail per method to execute two independent methods

### ISSUE THREAD

**Issue by bubahaze** [rieckpil/blog-tutorials#44]
## Blog post you are referring to

Refers to: blog post "Use GreenMail For Spring Mail (JavaMailSender) JUnit 5 Integration Tests"
https://rieckpil.de/use-greenmail-for-spring-mail-javamailsender-junit-5-integration-tests/
## Your Question

I test 2 methods inside test class, when I test them separately (withPerMethodLifecycle(false) -as in post) everything works fine, however if I test the whole class – the MimeMessage[] array in second method holds 2 messages (one from previous method I suppose), not one. Setting methodLifecycle to true makes loading application context impossible(“mail server is not available”, “failed to load ApplicationContext”). I did not understand what some code at blog post does, so I made some changes to undestand the code I'm using. Any ideas what might help? I’d be grateful!

```
@SpringBootTest
@ActiveProfiles("test")
@ExtendWith(MockitoExtension.class)
class EmailServiceTest {
    @RegisterExtension
    static GreenMailExtension greenMail = new GreenMailExtension(ServerSetupTest.SMTP)
            .withConfiguration(GreenMailConfiguration.aConfig().withUser("admin", "root"))
            .withPerMethodLifecycle(true);

    @Autowired
    EmailSender emailSender;
    final String RECIPIENT = "testuser@testing.com";
    final String CONTENT = "test content";
    final String SENDER = "test.sender@gmail.com";

    @Test
    void sendConfirmationEmail_should_send_confirmation_email() throws MessagingException, IOException {

        String subject = "Confirm your email";
        greenMail.start();

        emailSender.sendConfirmationEmail(RECIPIENT, CONTENT);
        MimeMessage[] receivedMessages = greenMail.getReceivedMessages();

        assertThat(receivedMessages.length).isEqualTo(1);

        await().atMost(2, SECONDS).untilAsserted(() -> {
            MimeMessage receivedMessage = receivedMessages[0];
            assertThat(receivedMessage.getAllRecipients()[0].toString()).isEqualTo(RECIPIENT);
            assertThat(receivedMessage.getFrom()[0].toString()).isEqualTo(SENDER);
            assertThat(receivedMessage.getSubject()).isEqualTo(subject);
            assertThat(receivedMessage.getContent().toString()).contains(CONTENT);
        });
    }

    @Test
    void sendNotificationEmail_should_send_notification_email() throws MessagingException, IOException {
        String subject = "Notification about recent changes to issue reported by you";
        emailSender.sendNotificationEmail(RECIPIENT, CONTENT);
        MimeMessage[] receivedMessages = greenMail.getReceivedMessages();
        await().atMost(2, SECONDS).untilAsserted(() -> {
            assertThat(receivedMessages.length).isEqualTo(1); 
            MimeMessage receivedMessage = receivedMessages[0];
            assertThat(receivedMessage.getAllRecipients()[0].toString()).isEqualTo(RECIPIENT);
            assertThat(receivedMessage.getFrom()[0].toString()).isEqualTo(SENDER);
            assertThat(receivedMessage.getSubject()).isEqualTo(subject);
            assertThat(receivedMessage.getContent().toString()).contains(CONTENT);
        });

    }
}
```
`application-test.yml`

```yaml
spring:
    mail:
        password: root
        username: admin
        host: localhost
        port: 3025 # default protocol port + 3000 as offset-->
        protocol: smtp
        test-connection: true
```

Stacktrace:
```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration': Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration]: Constructor threw exception; nested exception is java.lang.IllegalStateException: Mail server is not available
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:315) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:296) ~[spring-beans-5.3.13.jar:5.3.13]
...
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration]: Constructor threw exception; nested exception is java.lang.IllegalStateException: Mail server is not available
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:311) ~[spring-beans-5.3.13.jar:5.3.13]
	...
Caused by: java.lang.IllegalStateException: Mail server is not available
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.validateConnection(MailSenderValidatorAutoConfiguration.java:54) ~[spring-boot-autoconfigure-2.6.0.jar:2.6.0]
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.<init>(MailSenderValidatorAutoConfiguration.java:46) ~[spring-boot-autoconfigure-2.6.0.jar:2.6.0]
...
Caused by: java.net.ConnectException: Connection refused: connect
	at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method) ~[na:na]
	at java.base/java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:107) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224) ~[na:na]
	...

2022-02-04 10:46:51.827 ERROR 15320 --- [    Test worker] o.s.test.context.TestContextManager      : Caught exception while allowing TestExecutionListener [org.springframework.test.context.web.ServletTestExecutionListener@3088660d] to prepare test instance [com.poludnikiewicz.bugtracker.email.EmailServiceTest@6c98079]

java.lang.IllegalStateException: Failed to load ApplicationContext
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:132) ~[spring-test-5.3.13.jar:5.3.13]
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:124) ~[spring-test-5.3.13.jar:5.3.13]
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:190) ~[spring-test-5.3.13.jar:5.3.13]
...
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration]: Constructor threw exception; nested exception is java.lang.IllegalStateException: Mail server is not available
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:311) ~[spring-beans-5.3.13.jar:5.3.13]
	... 
Caused by: java.lang.IllegalStateException: Mail server is not available
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.validateConnection(MailSenderValidatorAutoConfiguration.java:54) ~[spring-boot-autoconfigure-2.6.0.jar:2.6.0]
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.<init>(MailSenderValidatorAutoConfiguration.java:46) ~[spring-boot-autoconfigure-2.6.0.jar:2.6.0]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:na]
...
Caused by: com.sun.mail.util.MailConnectException: Couldn't connect to host, port: localhost, 3025; timeout 5000
	at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2210) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at com.sun.mail.smtp.SMTPTransport.protocolConnect(SMTPTransport.java:722) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at javax.mail.Service.connect(Service.java:342) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at org.springframework.mail.javamail.JavaMailSenderImpl.connectTransport(JavaMailSenderImpl.java:518) ~[spring-context-support-5.3.13.jar:5.3.13]
...
Caused by: java.net.ConnectException: Connection refused: connect
	at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method) ~[na:na]
	at java.base/java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:107) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242) ~[na:na]
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224) ~[na:na]
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:na]
	at java.base/java.net.Socket.connect(Socket.java:609) ~[na:na]
	at com.sun.mail.util.WriteTimeoutSocket.connect(WriteTimeoutSocket.java:91) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at com.sun.mail.util.SocketFetcher.createSocket(SocketFetcher.java:333) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at com.sun.mail.util.SocketFetcher.getSocket(SocketFetcher.java:214) ~[jakarta.mail-1.6.7.jar:1.6.7]
	at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2160) ~[jakarta.mail-1.6.7.jar:1.6.7]
	...
Failed to load ApplicationContext
java.lang.IllegalStateException: Failed to load ApplicationContext
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:132)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:124)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:190)
...
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration': Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration]: Constructor threw exception; nested exception is java.lang.IllegalStateException: Mail server is not available
	at app//org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:315)
	at app//org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:296)
...
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration]: Constructor threw exception; nested exception is java.lang.IllegalStateException: Mail server is not available
	at app//org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224)
	at app//org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117)
	at app//org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:311)
	... 
Caused by: java.lang.IllegalStateException: Mail server is not available
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.validateConnection(MailSenderValidatorAutoConfiguration.java:54)
	at org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.<init>(MailSenderValidatorAutoConfiguration.java:46)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	...
Caused by: com.sun.mail.util.MailConnectException: Couldn't connect to host, port: localhost, 3025; timeout 5000;
  nested exception is:
	java.net.ConnectException: Connection refused: connect
	at app//com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2210)
	at app//com.sun.mail.smtp.SMTPTransport.protocolConnect(SMTPTransport.java:722)
	at app//javax.mail.Service.connect(Service.java:342)
	at app//org.springframework.mail.javamail.JavaMailSenderImpl.connectTransport(JavaMailSenderImpl.java:518)
	at app//org.springframework.mail.javamail.JavaMailSenderImpl.testConnection(JavaMailSenderImpl.java:398)
	at app//org.springframework.boot.autoconfigure.mail.MailSenderValidatorAutoConfiguration.validateConnection(MailSenderValidatorAutoConfiguration.java:51)
	... 
Caused by: java.net.ConnectException: Connection refused: connect
	at java.base/java.net.PlainSocketImpl.waitForConnect(Native Method)
	at java.base/java.net.PlainSocketImpl.socketConnect(PlainSocketImpl.java:107)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at com.sun.mail.util.WriteTimeoutSocket.connect(WriteTimeoutSocket.java:91)
	at com.sun.mail.util.SocketFetcher.createSocket(SocketFetcher.java:333)
	at com.sun.mail.util.SocketFetcher.getSocket(SocketFetcher.java:214)
	at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2160)
	...


```

## Context (Environment, Dependency Versions, Framework)

GreenMail, SpringBoot, JUnit5, Java 11

**Comment by rieckpil** [comment]
Thanks a lot for raising this PR @bubahaze. I just spent an entire hour fighting with it as I could reproduce this with my code example when using `.withPerMethodLifecycle(true)` and it was driving me crazy 🤯

The issue lies in the `application-test.yml` and `test-connection` set to `true`. When using a GreenMail server per-test-method approach, the GreenMail extension will start/stop the server within JUnit Jupiter's `beforeEach/afterEach` phase (see the `GreenMailExtension` for this).

Somehow the Spring Boot autoconfiguration for the `JavaMailSender` kicks in more early than GreenMail and hence upon Spring Boot initialization it tries to test the connection (`test-connection:true`) before GreenMail ever had the chance to start the server.

Long story short: Set `test-connection` to `false` and everything works as expected 🚀

**Comment by rieckpil** [comment]
Also, make sure to remove the manual `greenMail.start();` from your first test.

**Comment by bubahaze** [comment]
Wow, I would never think about the test-connection here. So glad you did! Thanks a lot for spending your time on this and efforts, really appreciate it. Cheers!

**Comment by rieckpil** [comment]
Closing this issue now 👍 Let me know in case it still fails on your end.
